File: Codigo/dindin-api/app/models/Transaction.js
const { Model, DataTypes } = require("sequelize");

const Category = require("./Category");
const TransactionRecurrencies = require("./TransactionRecurrencies");
const User = require("./User");
const Wallet = require("./Wallet");

class Transaction extends Model {
    static init(sequelize) {
        super.init(
            {
                id: {
                    field: "id",
                    type: DataTypes.UUID,
                    defaultValue: DataTypes.UUIDV4,
                    primaryKey: true,
                    unique: true,
                    allowNull: false,
                    required: true,
                    notEmpty: true,
                    comment: "UUIDV4",
                },
                wallet_id: {
                    field: "wallet_id",
                    type: DataTypes.BIGINT.UNSIGNED,
                    allowNull: false,
                    references: {
                        model: Wallet,
                        key: "id",
                    },
                },
                user_id: {
                    field: "user_id",
                    type: DataTypes.INTEGER.UNSIGNED,
                    allowNull: false,
                    references: {
                        model: User,
                        key: "id",
                    },
                },
                description: {
                    field: "description",
                    type: DataTypes.STRING(30),
                    allowNull: false,
                    notEmpty: true,
                },
                value: {
                    type: DataTypes.DOUBLE,
                    allowNull: false,
                },
                category_id: {
                    field: "category_id",
                    type: DataTypes.BIGINT.UNSIGNED,
                    allowNull: true,
                    references: {
                        model: Category,
                        key: "id",
                    },
                },
                transaction_recurrencies_id: {
                    field: "transaction_recurrencies_id",
                    type: DataTypes.BIGINT.UNSIGNED,
                    allowNull: true,
                    references: {
                        model: TransactionRecurrencies,
                        key: "id",
                    },
                },
                created_at: {
                    type: DataTypes.DATE,
                    allowNull: false,
                },
                updated_at: {
                    type: DataTypes.DATE,
                    allowNull: false,
                },
                deleted_at: {
                    type: DataTypes.DATE,
                    allowNull: true,
                    defaultValue: null,
                },
            },
            {
                tableName: "transaction",
                charset: "utf8mb4",
                collate: "utf8mb4_bin",
                timestamps: true, // deleted_at and updatedAt need this
                paranoid: true, // deleted_at need this
                createdAt: "created_at",
                updatedAt: "updated_at",
                deletedAt: "deleted_at", // .destroy() and .destroy(); to softdelete
                sequelize,
            }
        );
    }
}

module.exports = Transaction;

File: Codigo/dindin-mobile/dindin/lib/models/transaction.dart
import 'package:dindin/models/category.dart';
import 'package:dindin/models/transactionRecurrencies.dart';
import 'package:dindin/models/wallet.dart';

class Transaction {
  final String id;
  final num value;
  final Category? category;
  final TransactionRecurrencies? transactionRecurrencies;
  final String walletId;
  final String date;
  final String description;
  final String createdAt;

  const Transaction(
      {required this.value,
      required this.description,
      required this.createdAt,
      required this.id,
      required this.category,
      required this.date,
      required this.walletId,
      required this.transactionRecurrencies,
      });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'],
      value: json['value'],
      category:
          json['category'] != null ? Category.fromJson(json['category']) : null,
      description: json['description'],
      createdAt: json['created_at'],
      walletId: json['wallet_id'],
      date: json['date'],
      transactionRecurrencies: json['transaction_recurrencies'] != null ? TransactionRecurrencies.fromJson(json['transaction_recurrencies']) : null,
    );
  }
}
